packages/test-runner-intercept/src/interceptModulePlugin.ts

```diff
@@ -0,0 +1,84 @@
+import { Plugin } from '@web/dev-server-core';
+import { readFile } from 'fs/promises';
+import { parse } from 'es-module-lexer';
+import { createResolveImport, ResolveImport } from './createResolveImport';
+import { stripColor } from './stripColor';
+
+/**
+ * Plugin that allows the interception of modules
+ */
+export function interceptModulePlugin(): Plugin {
+  const paths: string[] = [];
+
+  let resolveImport: ResolveImport;
+  return {
+    name: 'intercept-module',
+
+    serverStart(params) {
+      resolveImport = createResolveImport(params, this);
+    },
+    async serve(context) {
+      if (context.path.startsWith('/__intercept__/')) {
+        let body;
+        try {
+          const source = context.path.replace('/__intercept__/', '');
+          paths.push(source);
+
+          const resolvedPath = await resolveImport({ context, source });
+          const url = new URL(resolvedPath as string, context.request.href);
+          const relativeFilePath = url.pathname.substring(1);
+
+          const content = await readFile(relativeFilePath, 'utf-8');
+          const [, exports] = await parse(content, resolvedPath as string);
+
+          const namedExports = exports.map(e => e.n).filter(n => n !== 'default');
+          const hasDefaultExport = exports.some(e => e.n === 'default');
+
+          body = `
+          import * as original from '${resolvedPath}';
+          const newOriginal = {...original};
+
+          ${namedExports.map(x => `export let ${x} = newOriginal['${x}'];`).join('\n')}
+
+          ${
+            hasDefaultExport
+              ? `
+          function computeDefault() {
+            if(typeof newOriginal.default === 'function'){
+              return (...args) => newOriginal.default.call(undefined, ...args);
+            }
+            return newOriginal.default;
+          }
+
+          export default computeDefault()
+          `
+              : ''
+          }
+
+          export const __wtr_intercepted_module__ = new Proxy(newOriginal, {
+            set: function(obj, prop, value) {
+              ${namedExports.map(x => `if (prop === '${x}') { ${x} = value;}`).join('\n')}
+              return Reflect.set(obj, prop, value);
+            },
+            defineProperty(target, key, descriptor) {
+              ${namedExports.map(x => `if (key === '${x}') { ${x} = descriptor.value;}`).join('\n')}
+              return Reflect.defineProperty(target, key, descriptor);
+            },
+          });
+        `;
+        } catch (error) {
+          body = `export const __wtr_error__ = '${stripColor((error as Error).message)}';`;
```

Comment:
what makes you need stripColor?

Reply:
Resolve plugins can use ANSI colors in their error messages (eg. https://github.com/modernweb-dev/web/blob/3a9d0073ac39dec75b01a3a422584706dd064d9e/packages/dev-server-rollup/src/rollupAdapter.ts#L247C90-L247C90). Unfortunately these color codes are only working in terminal output, and not in browser's console.

Maybe I should add a comment to detail this here?

Reply:
> Unfortunately these color codes are only working in terminal output, and not in browser's console.

don't they? I think DevTools console supports colors too
but thanks for explaining, not sure it's good idea to copy color even if it does, up to you

the only thing I'm concerned is that you need to implement this utility yourself, it doesn't seem like the most straightforward thing to me to maintain, but added a 3rd party dep to remove it is also not good, so you can keep things as is if you decided it's needed

Reply:
a comment might be good in this case indeed

Reply:
> don't they? I think DevTools console supports colors too

It does, but not using the terminal code convention.

>  not sure it's good idea to copy color even if it does, up to you

The idea is not to copy the color. :-) If I don't strip the codes, the error looks broken in the browser.
![image](https://github.com/modernweb-dev/web/assets/17364880/5ea99620-8a7c-400c-917c-38fdc291ca56)

> a comment might be good in this case indeed

Will add one.

Reply:
Done!